eclair-core/src/main/scala/fr/acinq/eclair/channel/publish/RawTxPublisher.scala

```diff
@@ -0,0 +1,190 @@
+
+import scala.concurrent.ExecutionContext
+import scala.concurrent.duration.DurationLong
+import scala.util.{Failure, Random, Success}
+
+/**
+ * Created by t-bast on 10/06/2021.
+ */
+
+/**
+ * This actor publishes a raw transaction without modifying it.
+ * It waits for confirmation or failure before reporting back to the requesting actor.
+ */
+object RawTxPublisher {
+
+  // @formatter:off
+  sealed trait Command
+  case class Publish(replyTo: ActorRef[TxPublisher.PublishTxResult], cmd: TxPublisher.PublishRawTx) extends Command
+  private case object TimeLocksOk extends Command
+  private case object CheckParentTx extends Command
+  private case object ParentTxOk extends Command
+  private case object ParentTxMissing extends Command
+  private case class WrappedTxResult(result: MempoolTxMonitor.TxResult) extends Command
+  private case class UnknownFailure(reason: Throwable) extends Command
+  case object Stop extends Command
+  // @formatter:on
+
+  def apply(nodeParams: NodeParams, bitcoinClient: ExtendedBitcoinClient, watcher: ActorRef[ZmqWatcher.Command], loggingInfo: TxPublishLogContext): Behavior[Command] = {
+    Behaviors.setup { context =>
+      Behaviors.withTimers { timers =>
+        Behaviors.withMdc(loggingInfo.mdc()) {
+          new RawTxPublisher(nodeParams, bitcoinClient, watcher, context, timers, loggingInfo).start()
+        }
+      }
+    }
+  }
+
+}
+
+private class RawTxPublisher(nodeParams: NodeParams,
+                             bitcoinClient: ExtendedBitcoinClient,
+                             watcher: ActorRef[ZmqWatcher.Command],
+                             context: ActorContext[RawTxPublisher.Command],
+                             timers: TimerScheduler[RawTxPublisher.Command],
+                             loggingInfo: TxPublishLogContext)(implicit ec: ExecutionContext = ExecutionContext.Implicits.global) {
+
+  import RawTxPublisher._
+
+  private val log = context.log
+
+  def start(): Behavior[Command] = {
+    Behaviors.receiveMessagePartial {
+      case Publish(replyTo, cmd) => checkTimeLocks(replyTo, cmd)
+      case Stop => Behaviors.stopped
+    }
+  }
+
+  def checkTimeLocks(replyTo: ActorRef[TxPublisher.PublishTxResult], cmd: TxPublisher.PublishRawTx): Behavior[Command] = {
+    val timeLocksChecker = context.spawn(TxTimeLocksMonitor(nodeParams, watcher, loggingInfo), "time-locks-monitor")
+    timeLocksChecker ! CheckTx(context.messageAdapter[TxTimeLocksMonitor.TimeLocksOk](_ => TimeLocksOk), cmd.tx, cmd.desc)
+    Behaviors.receiveMessagePartial {
+      case TimeLocksOk => checkParentPublished(replyTo, cmd)
+      case Stop =>
+        timeLocksChecker ! TxTimeLocksMonitor.Stop
+        Behaviors.stopped
+    }
+  }
+
+  def checkParentPublished(replyTo: ActorRef[TxPublisher.PublishTxResult], cmd: TxPublisher.PublishRawTx): Behavior[Command] = {
+    cmd.parentTx_opt match {
+      case Some(parentTx) =>
+        context.self ! CheckParentTx
```

Comment:
Why don't we always check the parent tx and retry after a delay? We could use the inputs of the tx to be published?

Reply:
It would be a bit wasteful and almost never necessary.
With anchor outputs we never need this because there's always a CSV 1.
It's only necessary for non-anchor htlc txs (and only because there's a race condition on publishing), but these should be gone in the not-so-distant future.

Reply:
Ok, but still I don't think we need to provide the full parent tx, since we only use the txid to make the check. If we always did the check (agreed, not optimal), we could simplify the API a little bit by removing parentTx_opt entirely. Not a big deal though.

Reply:
I changed it to only contain the parent `txid` in https://github.com/ACINQ/eclair/pull/1844/commits/eb4ddfcd362ad867feb5a3d1a198c27c69cf198d to make it lighter